diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -543,29 +552,117 @@ bool DifferentialFunctionComparator::mayIgnore(const User *Inst) const {
-                    if (isa<BinaryOperator>(U)) {
-                        return false;
-                    } else if (isa<CastInst>(U)) {
-                        for (const User *UU : U->users())
-                            UserStack.push_back(UU);
+            }
+        } else if (auto Load = dyn_cast<LoadInst>(Inst)) {
+            return mayIgnoreLoad(Load, addToIgnored);
+        }
+    }
+
+    if (canBeIgnored && addToIgnored) {
+        ignoredInstructions.insert({Inst, replacement});
+    }
+    return canBeIgnored;
+}
+
+/// Check whether the given instruction is a repetitive variant of a previous
+/// load with no store instructions in between. Will return the first load
+/// if the replacement pointer is given.
+bool DifferentialFunctionComparator::mayIgnoreLoad(const LoadInst *Load,
+                                                   bool addToIgnored) const {
+    if (auto BB = Load->getParent()) {
+        // Check if all predecessor blocks have a path to a load
+        // from the same pointer with no stores to the same pointer
+        // in between.
+        bool first = true;
+        const LoadInst *PreviousLoad = nullptr;
+        std::deque<const BasicBlock *> blockQueue = {BB};
+        std::vector<const BasicBlock *> visitedBlocks;
+        while (!blockQueue.empty()) {
+            BB = blockQueue.front();
+            blockQueue.pop_front();
+            visitedBlocks.push_back(BB);
+
+            bool searchPredecessors = true;
+            for (auto it = BB->rbegin(); it != BB->rend(); ++it) {
+                // Skip all instructions before the compared load
+                // when in the first block.
+                if (first) {
+                    if (&*it == Load) {
+                        first = false;
+                    }
+                    continue;
+                }
+
+                if (auto OrigLoad = dyn_cast<LoadInst>(&*it)) {
+                    // Try to find a previous load corresponding to the same
+                    // pointer. When found, end the seach for the current
+                    // control flow branch.
+                    if (Load->getPointerOperand()
+                        == OrigLoad->getPointerOperand()) {
+                        PreviousLoad = OrigLoad;
+                        searchPredecessors = false;
+                        break;
+                    }
+                } else if (auto Store = dyn_cast<StoreInst>(&*it)) {
+                    // Check whether a possibly conflicting store instruction
+                    // is present. If found, end the search with a failure.
+                    PreviousLoad = nullptr;
+                    searchPredecessors = false;
+                    blockQueue.clear();
+                    break;
+                }
+            }
+
+            if (searchPredecessors) {
+                auto BBPredecessors = predecessors(BB);
+                if (BBPredecessors.begin() == BBPredecessors.end()) {
```

Comment:
You can use `BBPredecessors.empty()`.

Reply:
Unfortunately, this is not possible. The iterator returned by `predecessors` implements only `begin` and `end`.

Reply:
Just to clarify this: `iterator_range::empty()` was added in LLVM 10, therefore it can't be used because of DiffKemp supporting lower versions of LLVM.